File: 23.MergekSortedLists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    # use count because:
    # it handles the case of a "tie" when two list nodes have the same value. When that happens, Python will look at the next value in the tuple (in this case, count), and sort based on that value. Without count, a "tie" would error out if the next value in the tuple were a ListNode (which can't be compared).
    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        pq = []
        dummy = ListNode(0)
        curr = dummy
        count = 0
        for node in lists:
            if node:
                count += 1
                heapq.heappush(pq, (node.val, count, node))
        while len(pq) > 0:
            currMin = heapq.heappop(pq)
            curr.next = currMin[2]
            curr = curr.next
            if curr.next:
                count += 1
                heapq.heappush(pq, (curr.next.val, count, curr.next))
        return dummy.next
   

File: 23.MergekSortedLists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        if (lists==null||lists.length==0) return null;
        
        PriorityQueue<ListNode> pq = new PriorityQueue<ListNode>(lists.length, new Comparator<ListNode>() {
            @Override
            public int compare(ListNode l1, ListNode l2) {
                if (l1.val < l2.val) return -1;
                else if (l1.val > l2.val) return 1;
                return 0;
            }
        });
        
        for (ListNode node : lists) {
            if (node != null)
                pq.offer(node);
        }
        
        ListNode dummy = new ListNode(0);
        ListNode tail = dummy;
        while (!pq.isEmpty()) {
            ListNode curr = pq.poll();
            tail.next = curr;
            tail = tail.next;
            if (curr.next != null) {
                pq.offer(curr.next);
            }
        }
        return dummy.next;
    }
}